### PR TITLE
Add helper functions in getMatchingCentralPaneRouteForState

### PR DESCRIPTION
**Pull request by WojtekBoman**
### Details

This PR fixes the issue that appears when we change the selected screen in the settings tab, switch to the chat tab and return to the settings. Then in this tab we should see the previously opened screen, but currently the Profile screen here is always displayed as the initial one.

### Fixed Issues

$ https://github.com/Expensify/App/issues/35613
PROPOSAL: N/A

### Tests

1. Open the settings tab.
2. Select any screen other than Profile on this tab.
3. Open the chat tab.
4. Go back to the settings tab and verify that the screen selected earlier is currently displayed.

- [ ] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as tests.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

This issue doesn't occur on the small screen size, so the videos are attached only for:
- `MacOS: Chrome / Safari`
- `MacOS: Desktop`

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47774969/b9820bcc-330c-4752-bdc2-21c4011795ca

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/47774969/e3c1ef90-9369-4f87-ad72-cb4df59fd3fc
</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
@cubuspl42 -> https://github.com/Expensify/App/issues/35613#issuecomment-1948600152
please review this PR when you have time 🙇

File: src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts [hunk 1]
```diff
@@ -31,8 +31,45 @@ const getTopMostReportIDFromRHP = (state: State): string => {
     return '';
 };
 
+// Function that checks if the given route has a policyID equal to the id provided in the function params
```
Comment (cubuspl42) [inline src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:34]: When commenting functions, we don't focus on the fact that they're functions. We either write "Do something such and such" or "Does something and something".

File: src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts [hunk 2]
```diff
@@ -31,8 +31,45 @@ const getTopMostReportIDFromRHP = (state: State): string => {
...
+// Function that restores already opened screen in the settings tab. Thanks to this function, we are able to open the already selected screen in the workspace settings when we return from another tab.
+function getAlreadyOpenedSettingsScreen(rootState?: State, policyID?: string): keyof CentralPaneNavigatorParamList | undefined {
+    if (!rootState) {
+        return undefined;
+    }
+
+    // If one of the screen from WORKSPACES_SCREENS is now in the navigation state, we can decide which screen we should display.
+    // A screen from the navigation state can be pushed to the navigation state again only if it has a matching policyID with the currently selected workspace.
+    // Otherwise, when we switch the workspace, we want to display the initial screen in the settings tab.
+    const WORKSPACES_SCREENS = TAB_TO_CENTRAL_PANE_MAPPING[SCREENS.WORKSPACE.INITIAL].concat(TAB_TO_CENTRAL_PANE_MAPPING[SCREENS.ALL_SETTINGS]);
```
Comment (cubuspl42) [inline src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:54]: All local variables in the project are named using lower camel case convention.

File: src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts [hunk 3]
```diff
@@ -31,8 +31,45 @@ const getTopMostReportIDFromRHP = (state: State): string => {
     return '';
 };
 
+// Function that checks if the given route has a policyID equal to the id provided in the function params
+function hasRouteMatchingPolicyID(route: NavigationPartialRoute<CentralPaneName>, policyID?: string) {
+    return (
+        route?.params &&
+        'params' in route?.params &&
+        route.params.params &&
+        'policyID' in (route.params.params as Record<string, string | undefined>) &&
```
Comment (cubuspl42) [inline src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:40]: Please don't type assert this twice, extract a local variable of desired target type.

File: src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts [hunk 4]
```diff
@@ -31,8 +31,45 @@ const getTopMostReportIDFromRHP = (state: State): string => {
...
+        return undefined;
+    }
+
+    // If one of the screen from WORKSPACES_SCREENS is now in the navigation state, we can decide which screen we should display.
+    // A screen from the navigation state can be pushed to the navigation state again only if it has a matching policyID with the currently selected workspace.
+    // Otherwise, when we switch the workspace, we want to display the initial screen in the settings tab.
+    const WORKSPACES_SCREENS = TAB_TO_CENTRAL_PANE_MAPPING[SCREENS.WORKSPACE.INITIAL].concat(TAB_TO_CENTRAL_PANE_MAPPING[SCREENS.ALL_SETTINGS]);
+
+    const alreadyOpenedSettingsTab = rootState.routes
+        .filter((item) => item.params && 'screen' in item.params && WORKSPACES_SCREENS.includes(item.params.screen as keyof CentralPaneNavigatorParamList))
```
Comment (cubuspl42) [inline src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:57]: There's a lot of type assertion going on (here and in other places). I gave it a quick look and I couldn't come up with any quick fixes. Do you have any ideas how we could at least reduce the number of type assertions (`something as SomeType`)?

**Review by jakub-trzebiatowski — CHANGES_REQUESTED** [review]

File: src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts (hunk 2 above)
Comment (cubuspl42) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:54]: But why is this local, if this doesn't depend on function arguments?

**Comment by cubuspl42** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/2590174/ebd307a1-8630-4589-88f8-1e743b7687be

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/2590174/0ad38d16-5929-4d9a-8450-3b9cfbf810d9

</details>

**Review by jakub-trzebiatowski — APPROVED** [review]
I also checked Native/mWeb, and they seem to be unaffected, as expected.

**Review by hayata-suenaga — APPROVED** [review]
thanks everyone for great work!

**Review by situchan — COMMENTED** [review]
Please consider this potential regression: https://github.com/Expensify/App/pull/36162#issuecomment-1961586890

**Comment by hayata-suenaga** [comment]
@cubuspl42 could you check if the regression situchan linked is reproducible and was caused by this PR?

File: src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts [hunk 5]
```diff
@@ -31,8 +33,47 @@ const getTopMostReportIDFromRHP = (state: State): string => {
     return '';
 };
 
+// Check if the given route has a policyID equal to the id provided in the function params
+function hasRouteMatchingPolicyID(route: NavigationPartialRoute<CentralPaneName>, policyID?: string) {
+    if (!route.params) {
+        return false;
+    }
+
+    const params = `params` in route?.params ? (route.params.params as Record<string, string | undefined>) : undefined;
```
[the remark this answers is not in the document]
Comment (cubuspl42) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:42]: I haven't reproduced it yet; I'm on it.

[the remark this answers is not in the document]
Comment (cubuspl42) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:42]: @WojtekBoman You can take a look at it, anyway. Unfortunately, it seems that it's very likely that the problem comes from this PR.

[the remark this answers is not in the document]
Comment (WojtekBoman) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:42]: I'll prepare a PR with a fix for it

[the remark this answers is not in the document]
Comment (hayata-suenaga) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:42]: @WojtekBoman thank you! please let us know once the PR is ready 🙇

[the remark this answers is not in the document]
Comment (situchan) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:42]: Let's try to fix before deploy to staging

[the remark this answers is not in the document]
Comment (hayata-suenaga) [reply src/libs/Navigation/linkingConfig/getMatchingCentralPaneRouteForState.ts:42]: the fix pr was merged 🎉 thank you for the quick fix Wojciech!

**Pull request merged**